File: apps/android/app/src/main/java/com/exponential/app/domain/WorkflowSelection.kt
package com.exponential.app.domain

/**
 * EXP-1084: the workflow page's PICKER model — pure, mirrored ×4 (web
 * `lib/workflow-selection.ts`, desktop `ui::workflow_view::Selection`, iOS
 * `WorkflowSelection.swift`) and locked by the contract fixture
 * `workflow-view.json` `selection`. The strip is the picker: `All` first
 * (position 0), then every node in DAG order ([order]). An EMPTY [ids] IS All.
 *
 * A click picks exactly that node (a click on the already picked chip keeps
 * it); a toggle adds or removes one node (the last one out = All); extend
 * picks the DAG-order range from the [anchor] to the node; a step moves ONE
 * position from the [cursor] (the last clicked or stepped node, else the last
 * picked one) through All + the nodes, clamped; prune drops nodes that left.
 */
data class WorkflowSelection(
    /** The picked nodes, in DAG order. Empty = All. */
    val ids: List<String> = emptyList(),
    /** Where an extend range starts: the last plain or toggling click. */
    val anchor: String? = null,
    /** The node a step moves from: the last one clicked or stepped to. */
    val cursor: String? = null,
) {
    val isAll: Boolean get() = ids.isEmpty()

    /** The ONE picked node, or null for All and multi-picks. */
    val single: String? get() = ids.singleOrNull()

    /** A plain click on a chip; null = the All chip, an unknown node = All. */
    fun click(id: String?, order: List<String>): WorkflowSelection {
        if (id == null || id !in order) return ALL
        return WorkflowSelection(listOf(id), id, id)
    }

    /** Add or remove one node, in DAG order; the last one out = All. */
    fun toggle(id: String, order: List<String>): WorkflowSelection {
        if (id !in order) return ALL
        val next = ids.toMutableSet()
        if (!next.remove(id)) next.add(id)
        val picked = order.filter { it in next }
        if (picked.isEmpty()) return ALL
        return WorkflowSelection(picked, id, id)
    }

    /** The DAG-order range from the anchor (else the node itself) to [id]. */
    fun extend(id: String, order: List<String>): WorkflowSelection {
        if (id !in order) return ALL
        val from = order.indexOf(anchor?.takeIf { it in order } ?: id)
        val to = order.indexOf(id)
        return WorkflowSelection(order.subList(minOf(from, to), maxOf(from, to) + 1).toList(), order[from], id)
    }

    /** The node a step moves from, or null on All. */
    private fun stepOrigin(): String? = when {
        ids.isEmpty() -> null
        cursor != null && cursor in ids -> cursor
        else -> ids.last()
    }

    /** The strip position a step moves from: All = 0, the first node = 1. */
    fun position(order: List<String>): Int {
        val origin = stepOrigin() ?: return 0
        val index = order.indexOf(origin)
        return if (index < 0) 0 else index + 1
    }

    /** One position through All + the nodes, clamped; lands on ONE node or All. */
    fun step(delta: Int, order: List<String>): WorkflowSelection {
        val next = (position(order) + delta).coerceIn(0, order.size)
        if (next == 0) return ALL
        val id = order[next - 1]
        return WorkflowSelection(listOf(id), id, id)
    }

    /** Drops nodes that left the workflow; none left = All. */
    fun prune(order: List<String>): WorkflowSelection {
        val kept = order.filter { it in ids }
        if (kept.size == ids.size) return this
        if (kept.isEmpty()) return ALL
        return WorkflowSelection(
            ids = kept,
            anchor = anchor?.takeIf { it in kept } ?: kept.first(),
            cursor = cursor?.takeIf { it in kept } ?: kept.first(),
        )
    }

    companion object {
        val ALL = WorkflowSelection()

        /** The strip's chip ids in DAG order — the order `nodeStrip` draws them. */
        fun order(strip: List<StripWave>): List<String> = strip.flatMap { wave -> wave.nodes.map { it.id } }
    }
}

File: apps/web/src/lib/workflow-selection.ts
// EXP-1084: the workflow page's PICKER model — pure, mirrored ×4 (desktop
// `ui::workflow_view::Selection`, iOS/Android `WorkflowSelection`) and locked
// by the contract fixture `workflow-view.json` `selection`. The node strip is
// the picker: `All` first (position 0), then every node in DAG order
// (`workflowNodeStrip`, waves left to right). A selection is a set of node
// ids; an EMPTY selection IS `All`.
//
// The rules: a click picks exactly that node (a click on the already picked
// chip keeps it); cmd/ctrl-click toggles one node in or out (the last one
// out = All); shift-click picks the DAG-order range from the anchor (the
// last plain or toggling click) to the node; a step moves ONE position from
// the cursor (the last clicked or stepped node, else the last picked one)
// through All + the nodes, clamped at both ends, landing on one node or All;
// pruning drops nodes that left the workflow (none left = All). A modified
// key never steps.

export interface StripSelection {
  /** The picked nodes, in DAG order. Empty = `All`. */
  ids: readonly string[]
  /** Where a shift-click range starts: the last plain or toggling click. */
  anchor: string | null
  /** The node the keyboard steps from: the last one clicked or stepped to. */
  cursor: string | null
}

export const ALL_SELECTION: StripSelection = { ids: [], anchor: null, cursor: null }

export interface SelectModifiers {
  /** Cmd/Ctrl-click: add or remove one node. */
  toggle?: boolean
  /** Shift-click: the range from the anchor to this node, in DAG order. */
  extend?: boolean
}

function inOrder(ids: Iterable<string>, order: readonly string[]): string[] {
  const picked = new Set(ids)
  return order.filter((id) => picked.has(id))
}

/** A click on a node chip. `null` = the `All` chip. */
export function selectNode(
  current: StripSelection,
  id: string | null,
  order: readonly string[],
  modifiers: SelectModifiers = {}
): StripSelection {
  if (id === null || !order.includes(id)) return ALL_SELECTION
  if (modifiers.extend) {
    const anchor =
      current.anchor && order.includes(current.anchor) ? current.anchor : id
    const from = order.indexOf(anchor)
    const to = order.indexOf(id)
    const range = order.slice(Math.min(from, to), Math.max(from, to) + 1)
    return { ids: range, anchor, cursor: id }
  }
  if (modifiers.toggle) {
    const next = new Set(current.ids)
    if (!next.delete(id)) next.add(id)
    const ids = inOrder(next, order)
    if (ids.length === 0) return ALL_SELECTION
    return { ids, anchor: id, cursor: id }
  }
  return { ids: [id], anchor: id, cursor: id }
}

/** ←/→ (and j/k): one step through `All` + the nodes in DAG order, landing on
 *  exactly ONE node (or `All`). Clamped at both ends. The face is the
 *  caller's and is never touched. */
export function stepSelection(
  current: StripSelection,
  direction: 1 | -1,
  order: readonly string[]
): StripSelection {
  const cursor =
    current.ids.length === 0
      ? null
      : current.cursor && current.ids.includes(current.cursor)
        ? current.cursor
        : current.ids[current.ids.length - 1]!
  const position = cursor === null ? 0 : order.indexOf(cursor) + 1
  const next = Math.min(Math.max(position + direction, 0), order.length)
  if (next === 0) return ALL_SELECTION
  const id = order[next - 1]!
  return { ids: [id], anchor: id, cursor: id }
}

/** Drops nodes that left the workflow (a replan folded them away). */
export function pruneSelection(
  current: StripSelection,
  order: readonly string[]
): StripSelection {
  const ids = inOrder(current.ids, order)
  if (ids.length === current.ids.length) return current
  if (ids.length === 0) return ALL_SELECTION
  return {
    ids,
    anchor: current.anchor && ids.includes(current.anchor) ? current.anchor : ids[0]!,
    cursor: current.cursor && ids.includes(current.cursor) ? current.cursor : ids[0]!,
  }
}

